### PR TITLE
Add support for multiple attribute group by

### PR DESCRIPTION
**Pull request by LahiruJayasekara**
- add multiple attribute group by and
                        - support user defined functions in group by
 in ballerina stream processing.

**Multiple attribute group by**
```
forever {
        from inputStream where inputStream.age > 25
        select inputStream.name, inputStream.age, count() as count
        group by inputStream.age, inputStream.school
        => (TeacherOutput [] teachers) {
            foreach t in teachers {
                outputStream.publish(t);
            }
        }
    }
```
**Using user defined group by functions**
```
function getGroupByField(int a) returns boolean {
    if (a % 2 == 0) {
        return true;
    } else {
        return false;
    }
}

forever {
        from inputStream where inputStream.age > 25
        select inputStream.name, inputStream.age, count() as count
        group by getGroupByField(inputStream.age)
        => (TeacherOutput [] teachers) {
            foreach t in teachers {
                outputStream.publish(t);
            }
        }
    }
```

## Assignees
@mohanvive , @gimantha , @grainier

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java [hunk 1]
```diff
@@ -840,23 +889,74 @@ private BLangLambdaFunction createAggregatorLambdaWithParams(BLangVariable varAg
...
+            addReturnGroupByFieldStmt(groupByLambda, conversionExpr);
+
+            groupByFunctionArray.add(groupingLambda);
+        }
+        return groupByFunctionArray;
+    }
+
+    private BLangInvocation refactorInvocationInGroupBy(BLangInvocation invocation,
+                                                        BLangVariable varGroupByStreamEvent) {
+        if (invocation.requiredArgs.size() > 0) {
```
Comment (NatashaWso2) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java]: Without checking if size of the required args is greater than zero at the start, shall we simply return the invocation by checking if size of the args is equal to zero. We can then place the code inside the if condition outside which will improve its clarity. WDYT?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java [hunk 2]
```diff
@@ -626,12 +632,34 @@ private void createSelectStatementWithGroupBy(BLangSelectClause selectClause) {
...
+
+        //create RHS expression for the group by function array
+        BLangArrayLiteral groupByFuncArrExpr = (BLangArrayLiteral) TreeBuilder.createArrayLiteralNode();
+        groupByFuncArrExpr.exprs = new ArrayList<>();
+        groupByFuncArrExpr.type = groupByFuncArrayType;
+
+        BLangVariableDef groupByFuncArrayVarDef =
+                createGroupByFuncArrayVarDef(selectClause, groupByFuncArrayVarSymbol, groupByFuncArrExpr);
+        stmts.add(groupByFuncArrayVarDef);
+
```
[the remark this answers is not in the document]
Comment (LahiruJayasekara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java]: fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java [hunk 3]
```diff
@@ -626,12 +632,34 @@ private void createSelectStatementWithGroupBy(BLangSelectClause selectClause) {
...
+        groupByFuncArrExpr.exprs = createGroupByLambda(selectClause);
+
```
[the remark this answers is not in the document]
Comment (LahiruJayasekara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java]: fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java [hunk 4]
```diff
@@ -840,23 +889,74 @@ private BLangLambdaFunction createAggregatorLambdaWithParams(BLangVariable varAg
...
+            BLangBlockStmt groupByLambda = groupingLambda.function.body;
+
+            //e.data; / getOrderByField(...);
+            BLangTypeConversionExpr conversionExpr = null;
+            if (node.getKind() == NodeKind.FIELD_BASED_ACCESS_EXPR) {
+                BLangIndexBasedAccess mapFieldAccessExpr = createMapAccessExprFromStreamEvent(varGroupByStreamEvent,
+                                                                                              (BLangExpression) node);
+                // return check <string>e.data[<fieldName in string>];
+                conversionExpr = generateConversionExpr(mapFieldAccessExpr, symTable.stringType, symResolver);
+
```
[the remark this answers is not in the document]
Comment (LahiruJayasekara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java]: fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java (hunk 1 above)
Comment (LahiruJayasekara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java]: That's good. Fixed

**Pull request merged**